### ISSUE TITLE
Introduce `trait NetworkValidationUnchecked {}`?

### ISSUE THREAD

**Issue by dpc** [rust-bitcoin/rust-bitcoin#3760]
So, I was brainstorming with @tvolk131 how to handle  `V: NetworkValidation` in Fedimint, and thinking how nice it would be if we could just parametrize structs using `Address` over `V: NetworkValidation` like:

```rust
/// Default is untrusted, to avoid mistakes by omission, need to validate externally provided input
#[derive(Serialize, Deserialize)]
struct FedimintSomeDTOType<V = NetworkUnchecked> {
  // ..
  address: Address<V>
  // ..
}

// for using in databases and other places that we know, have already all addresses checked
type FedimintSomeDTOTypeTrusted = FedimintSomeDTOType<V = NetworkChecked>;
```

This does not work because `Deserialize` is only `impl`-ed for `Address<NetworkUnchecked>`, rightly so, IMO.

The point here is that would be great if everything could be generic still, yet `Deserialize` could be derived, but still only for `<NetworkUnchecked>` versions.

I tried to use existing serde features to make it work, but was unable so I filled a serde feature request to support it: https://github.com/serde-rs/serde/issues/2869

But then I realized... if `rust-bitcoin` had the `trait NetworkValidationUnchecked : Network Validation {}` and then `impl NetworkValidationUnchecked for NetworkUnchecked` and then `impl Desierialize for Address<U> where U :  NetworkValidationUnchecked` then this mechanism would work even without extra serde features! We could just `#[serde(bound(deserialize = "V: NetworkValidateUnchecked"))]` in our code and it would pass. A little bit tricky to come up with, but seems worthwhile to allow cutting down on boilerplate.

And in general all other downstream consumers could use `where U :  NetworkValidationUnchecked` to make their own `impl`s work the same way. 

On top of my head I don't know if this has any downsides, and I came up with it just like 20 minutes ago, but reporting for debate.

**Comment by tcharding** [comment]
I had a go, stumbled on either inner/outer generics if I used Address<NetworkUnchecked> in the `Visitory` or missing `FromStr` if I used `U` (bolew) because we intentionally don't implement it.

```rust

/// Should be sealed too.
pub trait NetworkValidationUnchecked: Sync + Send + Sized + Unpin {}

impl NetworkValidationUnchecked for NetworkUnchecked {}

impl<'de, U: NetworkValidationUnchecked + NetworkValidation> serde::Deserialize<'de> for Address<U> {
    fn deserialize<D>(deserializer: D) -> Result<Address<U>, D::Error>
    where
        D: serde::de::Deserializer<'de>,
    {
        use core::fmt::Formatter;

        struct Visitor<U>(PhantomData<U>);
        impl<'de, U> serde::de::Visitor<'de> for Visitor<U>
        where U: NetworkValidationUnchecked + NetworkValidation,
              Address<U>: FromStr,
        {
            type Value = Address<U>;

            fn expecting(&self, f: &mut Formatter) -> core::fmt::Result {
                f.write_str("A Bitcoin address")
            }

            fn visit_str<E>(self, v: &str) -> core::result::Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                v.parse::<Address<U>>().map_err(E::custom)
            }
        }

        deserializer.deserialize_str(Visitor(PhantomData::<U>))
    }
}
```

```rust

    #[test]
    #[cfg(feature = "serde")]
    fn improved_serde() {
        use serde::{Serialize, Deserialize};

        #[derive(Serialize, Deserialize)]
        struct Foo<V>
            where V: NetworkValidation,
        {
            address: Address<V>,
        }
    }
```

```bash
error[E0277]: the trait bound `address::Address<U>: FromStr` is not satisfied
   --> bitcoin/src/address/mod.rs:170:22
    |
170 |         deserializer.deserialize_str(Visitor(PhantomData::<U>))
    |                      ^^^^^^^^^^^^^^^ the trait `FromStr` is not implemented for `address::Address<U>`, which is required by `<address::Address<U> as serde::Deserialize<'de>>::deserialize::Visitor<U>: serde::de::Visitor<'de>`
```

**Comment by dpc** [comment]
@tcharding Yes, that's the same problem. `FromStr` ewould need to use `U : NetworkValidationUnchecked` as well.

**Comment by dpc** [comment]
@tcharding In your test, `Foo` needs #[serde(bound(deserialize = "V: NetworkValidateUnchecked"))] as per https://serde.rs/container-attrs.html#bound

**Comment by apoelstra** [comment]
BTW, if you can come up with an analogous thing in rust-miniscript, even if it's more complicated, we might be more willing to experiment there.

I support this use-case and would like to see a simple solution ... but there's a limit to how much generic-propagation we can tolerate because it slows down compile times and makes the docs much harder to follow.

**Comment by dpc** [comment]
> BTW, if you can come up with an analogous thing in rust-miniscript, even if it's more complicated, we might be more willing to experiment there.

Can I do it there, without having it here? The whole problem is that once `Address<Unchecked>` is used it can not be turned into `Address<V>` downstream, because potentially `rust-bitcoin` could add any other type to impl `V` in the future.

> I support this use-case and would like to see a simple solution ... but there's a limit to how much generic-propagation we can tolerate because it slows down compile times and makes the docs much harder to follow.

I don't think this will change compilation time (much). There will always be only one monomorphisation of `deserialize` for `Address`, and any other `deserialize` downstream. The `Address` itself is already generic. Other methods would get monorphised usually once too, sometimes twice (if one is using both checked and unchecked version of the type).

**Comment by apoelstra** [comment]
> I don't think this will change compilation time (much). There will always be only one monomorphisation of `deserialize` for `Address`, and any other `deserialize` downstream. The `Address` itself is already generic. Other methods would get monorphised usually once too, sometimes twice (if one is using both checked and unchecked version of the type).

Yeah, on further thought I think you're right. I think adding the trait bound strictly simplifies things and the extra compilation cost will be trivial.

**Comment by dpc** [comment]
Asking around for 3rd opinions: https://users.rust-lang.org/t/using-trait-bound-instead-of-a-concrete-session-type/122610
